osu.Game/Online/API/APIRequest.cs

```diff
@@ -16,20 +17,27 @@ public abstract class APIRequest<T> : APIRequest where T : class
     {
         protected override WebRequest CreateWebRequest() => new OsuJsonWebRequest<T>(Uri);
 
-        public T Result => ((OsuJsonWebRequest<T>)WebRequest)?.ResponseObject;
+        public T Result { get; private set; }
 
         protected APIRequest()
         {
-            base.Success += onSuccess;
+            base.Success += () => TriggerSuccess(((OsuJsonWebRequest<T>)WebRequest)?.ResponseObject);
         }
 
-        private void onSuccess() => Success?.Invoke(Result);
-
         /// <summary>
         /// Invoked on successful completion of an API request.
         /// This will be scheduled to the API's internal scheduler (run on update thread automatically).
         /// </summary>
         public new event APISuccessHandler<T> Success;
+
+        internal void TriggerSuccess(T result)
+        {
+            // disallow calling twice
+            Debug.Assert(Result == null);
```

Comment:
I think this should be an exception instead. It's `internal` sure, but it's still much too open imo.